core/src/solution_submission.rs

```diff
@@ -238,15 +259,40 @@ async fn retry_with_gas_price_increase(
         } else {
             Some(BLOCK_TIMEOUT)
         };
-        let result = submit_solution(gas_price, block_timeout).await;
-        if !is_confirm_timeout(&result) || is_last_iteration {
-            return result;
+        let solution_future = submit_solution(gas_price, block_timeout);
+        let deadline_future = wait_for_deadline();
+        futures::pin_mut!(deadline_future);
```

Comment:
Why is this needed?

Reply:
`select` requires the future to be pinned (or unpin). Afaik this is the normal way to accomplish this. It's not needed for `solution_future` because that future happens to already be unpin.

Reply:
I tried to figure out why exactly the deadline future is not unpin, but wasn't able to (might have something to do with `pin_project!` wrapper on the Timeout struct). Do you have an explanation?

Reply:
With medium confidence I believe that that async functions simply do not propagate Unpin. The return type of `wait_for_deadline` is `impl Future` so even if we did nothing in its body it would still be treated as Unpin. The reason that `solution_future` is unpin is because it isn't an `impl Future` but a `BoxFuture`.

Reply:
> With medium confidence I believe that that async functions simply do not propagate Unpin

Correct, `async` functions are marked as `!Unpin` regardless of their body. For example even the following `async` function is `!Unpin`:
```rust
async fn answer() -> i32 { 42 }
```
In fact, an `async` functions get wrapped by the `const fn std::future::from_generator<T>(gen: T) -> impl Future<Output = T::Return>`, and their underlying type is explicitly `!Unpin`: 
https://github.com/rust-lang/rust/blob/8aa18cbdc5d4bc33bd61e2d9a4b643d87f5d21de/src/libcore/future/mod.rs#L64

> is unpin is because it isn't an impl Future but a BoxFuture

You could also `.boxed()` instead of `futures::pin_mut!`. You might also be able to `.fuse()` the future as well which I thought was needed for `select!`.